src/context.rs

```diff
@@ -0,0 +1,57 @@
+use std::{
+    path::Path,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use tree_sitter::Language;
+
+use crate::{rule::ResolvedRule, violation::Violation};
+
+pub struct Context {
```

Comment:
I was somewhat mimicking ESLint's [rule-authoring API](https://eslint.org/docs/latest/extend/custom-rules)

But the concept of "context" being used here is a little fuzzy at the moment

In ESLint's API, they are calling the rule's `create()` callback once for each file that the rule is being run against, so it supplies a `context` that is specific to that file

(which I guess would enable eg tailoring that rule/its internal state to the specific file that it's being run against?)

The big things that seems questionable about mimicking that here is that we need to parse tree-sitter queries for the rule "listeners", and so doing that parsing n times (where n = the # of files we're checking) seems like not the way to do things

It's possible that that query-parsing could be eg memoized (for the common case that for each file you're registering the same exact set of tree-sitter queries for the listeners) if we did in fact want to have something like the ESLint API's "scope" (where you can basically configure the rule once per file)?

But no idea what actually makes sense

So for the moment there's this "global to the linter run"-level `Context`, and then a `QueryMatchContext` that gets supplied to your listener callback

Reply:
Yup I haven't really "wired up" any "rule-driven language" stuff yet but that definitely seems right to me